src/tools/generate/generate_figure_images.js

```diff
@@ -0,0 +1,110 @@
+const fs = require('fs-extra');
+const puppeteer = require('puppeteer');
+
+const { find_markdown_files } = require('./shared');
+
+const take_single_screenshot = async (graphUrl, filename) => {
+  const browser = await puppeteer.launch();
+  const page = await browser.newPage();
+  await page.setViewport({
+    "width": 1280,
+    "height": 800,
+    "deviceScaleFactor": 2
+  });
+  await page.goto(graphUrl, {
+    waitUntil: 'networkidle2',
+  });
+  const el = await page.$('#embed_chart');
+  await el.screenshot({ path: filename });
+  await browser.close();
+}
+
+const generate_images = async (chapter_match) => {
+
+  if (chapter_match) {
+    // Remove any trailing .md and replace all paths with brackets to capture components
+    // en/2019/javascript.md -> (en)/(2019)/(javascript).md
+    chapter_match = chapter_match.replace(/\.md$/,'');
+    chapter_match = chapter_match.replace(/^content[\/\\]*/,'');
+    chapter_match = (process.platform != 'win32')
+                ? 'content\/' +  '(' + chapter_match.replace(/\//g,')/(') + ').md'
+                : 'content\\\\' +  '(' + chapter_match.replace(/\\/g,')\\\\(') + ').md';
+    re = new RegExp(chapter_match);
+  } else {
+    console.log('Please provide an argument of the form: en/2020/performance');
+    process.exit(1);
+  }
+
+  for (const file of await find_markdown_files()) {
+
+    let path, language, year, chapter;
+
+    try {
+      [path, language, year, chapter] = file.match(re);
+    } catch(error) {
+      // No match - skip to next in for loop
+      continue;
+    }
+
+    if (language !== "en") {
+      console.log("Skipping non-English chapter");
+      continue;
+    }
+
+    console.log(`Generating for the ${year} ${chapter} chapter:`);
+
+    const markdown = await fs.readFile(file, 'utf-8');
+
+    // Let's not depend on the order of arguments so test for both
+    const figure_regexp1 = /{{ figure_markup\([^}]*image=["']([^'"]*)["'][^}]*chart_url=["']([^'"]*)["'][^}]*\)[^}]*}}/g;
```

Comment:
maybe be more lenient in matching (e.g. `/{{\s*figure_markup`)

would it make sense to make it into a single regex? something like:
`/{{ figure_markup\([^}]*(image|chart_url)=["']([^'"]*)["'][^}]*(image|chart_url)=["']([^'"]*)["'][^}]*\)[^}]*}}/g` (assuming the property is not defined twice)

Reply:
Literally just pushed the whitespace change!

Good call on the single regex.

Reply:
Ah yes remembered why it can't be the one regex, it's cause I need to swap around the variables:

```js
        let image_file = match[1];
        let chart_url = match[2];

        if (figure_regexp == figure_regexp2) {
          image_file = match[2];
          chart_url = match[1];
        }
```

Reply:
Ah can just capture the param though and test that. Fixed in https://github.com/HTTPArchive/almanac.httparchive.org/pull/2461/commits/200fd722d7325e208d4910ed8b4e9074dabed0f1

Thanks - that makes it cleaner.